doc/python-type-hinting.md

```diff
@@ -80,8 +78,41 @@ This import is not necessary in Python 3.10.
 
 https://www.python.org/dev/peps/pep-0563/
 
+## Type Ignore
+
+To mark portions of the program that should not be covered by type hinting, use the following on a particular line:
+
+```python
+# type: ignore
+```
+
+A `# type: ignore` comment on a line by itself at the top of a file silences all errors in the file.
+
+`# type: ignore` should only be used when unavoidable. Ensure that a comment is added to explain why it has been used.
+
+## ParamSpec
+
+Used to forward the parameter types of one callable to another callable e.g. to add basic logging to a function, create a decorator add_logging to log function calls:
+
+```python
+T = TypeVar('T')
```

Comment:
More of a question to all the reviewers. As I mentioned in tech session, do we want to restrict types in TypeVar? Some explanation:

<img width="801" alt="Screenshot 2023-02-13 at 09 59 31" src="https://user-images.githubusercontent.com/53475968/218707702-6bc0b902-0f38-4c95-9440-1499a7b15824.png">

Reply:
Also do we want to enforce [NoReturn](https://docs.python.org/3/library/typing.html#typing.NoReturn) instead of `None`? It will be replaced by [Never](https://docs.python.org/3/library/typing.html#typing.Never) in python 3.11.

Reply:
As discussed, `NoReturn` isn't a replacement for `None`.

Reply:
I am in favour of encouraging TypeVar restricting where possible, as it improves the code documentation and error prevention.
However I think there might be legitimate cases for an unrestricted TypeVar, typically a very generic util function.

Reply:
New section now added for `TypeVar` with examples on how to implement it.